src/utility/svd.jl

```diff
@@ -0,0 +1,240 @@
+using TensorKit:
+    SectorDict,
+    _tsvd!,
+    _empty_svdtensors,
+    _compute_svddata!,
+    _create_svdtensors,
+    NoTruncation,
+    TruncationSpace
+CRCExt = Base.get_extension(KrylovKit, :KrylovKitChainRulesCoreExt)
+
+"""
+    PEPSKit.tsvd(t::AbstractTensorMap, alg; trunc=notrunc(), p=2)
+
+Wrapper around `TensorKit.tsvd` which dispatches on the `alg` argument.
+This is needed since a custom adjoint for `PEPSKit.tsvd` may be defined,
+depending on the algorithm. E.g., for `IterSVD` the adjoint for a truncated
+SVD from `KrylovKit.svdsolve` is used.
+"""
+function PEPSKit.tsvd(
+    t::AbstractTensorMap, alg; trunc::TruncationScheme=notrunc(), p::Real=2
+)
+    return TensorKit.tsvd(t; alg, trunc, p)
+end
+
+# Wrapper around Krylov Kit's GKL iterative SVD solver
+@kwdef struct IterSVD
+    alg::KrylovKit.GKL = KrylovKit.GKL(; tol=1e-14, krylovdim=25)
+    lorentz_broad::Float64 = 0.0
+    alg_rrule::Union{GMRES,BiCGStab,Arnoldi} = GMRES(; tol=1e-14)
+end
+
+# Compute SVD data block-wise using KrylovKit algorithm
+function TensorKit._tsvd!(
+    t, alg::IterSVD, trunc::Union{NoTruncation,TruncationSpace}, p::Real=2
+)
+    # early return
+    if isempty(blocksectors(t))
+        truncerr = zero(real(scalartype(t)))
+        return _empty_svdtensors(t)..., truncerr
+    end
+
+    Udata, Σdata, Vdata, dims = _compute_svddata!(t, alg, trunc)
+    U, S, V = _create_svdtensors(t, Udata, Σdata, Vdata, spacetype(t)(dims))
+    truncerr = trunc isa NoTruncation ? abs(zero(scalartype(t))) : norm(U * S * V - t, p)
+
+    return U, S, V, truncerr
+end
+function TensorKit._compute_svddata!(
+    t::TensorMap, alg::IterSVD, trunc::Union{NoTruncation,TruncationSpace}
+)
+    InnerProductStyle(t) === EuclideanProduct() || throw_invalid_innerproduct(:tsvd!)
+    I = sectortype(t)
+    A = storagetype(t)
+    Udata = SectorDict{I,A}()
+    Vdata = SectorDict{I,A}()
+    dims = SectorDict{I,Int}()
+    local Sdata
+    for (c, b) in blocks(t)
+        x₀ = randn(eltype(b), size(b, 1))
+        howmany = trunc isa NoTruncation ? minimum(size(b)) : blockdim(trunc.space, c)
+        S, lvecs, rvecs, info = KrylovKit.svdsolve(b, x₀, howmany, :LR, alg.alg)
+        if info.converged < howmany  # Fall back to dense SVD if not properly converged
+            U, S, V = TensorKit.MatrixAlgebra.svd!(b, TensorKit.SVD())
+            Udata[c] = U
+            Vdata[c] = V
+        else  # Slice in case more values were converged than requested
+            Udata[c] = stack(lvecs[1:howmany])
+            Vdata[c] = stack(rvecs[1:howmany])'
```

Comment:
```suggestion
            Udata[c] = stack(view(lvecs, 1:howmany))
            Vdata[c] = stack(adjoint, view(rvecs, 1:howmany); dims=1)
```
I think using the views removes one allocation. The implementation for Vdata is untested, and might remove another allocation, but you should probably check this.

Reply:
The views are definitely removing one allocation . The problem with the `stack(adjoint, ...)` construction is that a singleton dimension is created and you have to get rid of it, which again creates allocations. Consider for example

```julia
julia> v = [[1, 2, 3], [4, 5, 6]]
2-element Vector{Vector{Int64}}:
 [1, 2, 3]
 [4, 5, 6]

julia> @time stack(adjoint, v; dims=2)
  0.000009 seconds (1 allocation: 112 bytes)
1×2×3 Array{Int64, 3}:
[:, :, 1] =
 1  4

[:, :, 2] =
 2  5

[:, :, 3] =
 3  6

julia> @time dropdims(stack(adjoint, v; dims=2); dims=1)
  0.000010 seconds (4 allocations: 240 bytes)
2×3 Matrix{Int64}:
 1  2  3
 4  5  6

julia> @time @view stack(adjoint, v; dims=2)[1, :, :]
  0.000013 seconds (3 allocations: 192 bytes)
2×3 view(::Array{Int64, 3}, 1, :, :) with eltype Int64:
 1  2  3
 4  5  6

julia> @time stack(v)'
  0.000007 seconds (2 allocations: 128 bytes)
2×3 adjoint(::Matrix{Int64}) with eltype Int64:
 1  2  3
 4  5  6
```

I don't know whether there is an allocation-free way to get rid of the singleton dimension, so for now I will just use the outer adjoint `stack(view(rvecs, 1:howmany))'`.

Reply:
`stack(conj, view(rvecs, 1:howmany))` might do the trick

Reply:
also keep in mind that the last benchmark is not entirely fair, `adjoint(::Matrix{Int64})` is just a lazy adjoint implementation, so in principle you'd have to use `copy(stack(v)')` to instantiate the actual adjoint. This is also what is happening implicitly upon assignment: `Vdata[c] = V'` first creates the lazy wrapper, and then the `setindex!` converts it to the `valtype` of `Vdata`, which needs to instantiate it. 
The cleaner approach here is probably to construct `V'` and then return an `AdjointTensorMap`, but if this does not show up in a profiler we should probably not worry too much about it.
I would be happier to include an explicit `copy` statement though, instead of the implicit conversion during the `setindex!`, just to remind ourselves that this is happening

Reply:
Yes that's a fair point, I ignored the lazy behavior. I think for now `stack(conj, view(rvecs, 1:howmany))` is a good working solution. In that case an explicit `copy` statement is not needed, right? (i.e. there should be no hidden `copy` happening)